docs: add feegrant module tutorial (#949)
cosmos/sdk-tutorials: feegrant/index.md

---
parent:
  title: Understanding the Feegrant Module
order: 0
description: Use the Cosmos SDK feegrant module to grant the use of tokens to pay for fees from one account (the granter) to another account (the grantee).
---

> # Fee Grant Module

Reviewer: ```suggestion
# Feegrant Module
```

The [`feegrant`](https://docs.cosmos.network/v0.44/modules/feegrant/) module enables the granter (a user, contract, or module) to pay the fees for someone else (the grantee) when the grantee wants to broadcast a transaction on the blockchain. The granter retains full access to their tokens and is able to revoke the allowance at any time.

> ## Use Fee Grant to Grant an Allowance

Reviewer: ```suggestion
## Use Feegrant to Grant an Allowance
```

---
parent:
  title: Understanding the Fee Grant Module
order: 0

> description: Use the Cosmos SDK feegrant module to grant the use of tokens to pay for fees from one account (the granter) to another account (the grantee).

Reviewer 2: ```suggestion
description: Use the Cosmos SDK fee grant module to grant the use of tokens to pay for fees from one account (the granter) to another account (the grantee).
```

---

> # Feegrant Module
> 
> The [feegrant](https://docs.cosmos.network/v0.44/modules/feegrant/) enables a user / contract / module (the granter) to pay the fees for someone else (the grantee) when they want to broadcast a transaction on the blockchain. A key feature of feegrant is that the granter retains full access to their tokens, and is able to revoke the allowance at any time.

Reviewer 2: ```suggestion
# Fee Grant Module

The [fee grant](https://docs.cosmos.network/v0.44/modules/feegrant/) module enables the granted (a user, contract, or module) to pay the fees for someone else (the grantee) when the grantee wants to broadcast a transaction on the blockchain. A key feature of fee grant is for the granter to retain full access to their tokens and be able to revoke the allowance at any time.
```

Reviewer 2: Several points here:

- fee grant = two words (I didn't mark the rest, please update all)
- avoid "they" and instead repeat the subject (I guessed at "the grantee" for "they" in this suggested update)
- do not use slashes to separate alternatives, see https://developers.google.com/style/slashes?hl=en

Reviewer 2: @nooomski I was trying to repair a painful sentence... Let's try this 

> The granter retains full access to their tokens and is able to revoke the allowance at any time.

Reviewer 2: and yes "granted" was my human error, thanks for catching that

## Use Feegrant to Grant an Allowance

An often discussed use case for the feegrant module is that blockchains or smartcontracts can allow new users to start interacting with the chain or contract without having to acquire tokens first, greatly improving the onboarding experience.

> There are two [fee allowance types](https://docs.cosmos.network/v0.44/modules/feegrant/01_concepts.html#fee-allowance-types) that the freegrant module implements: `BasicAllowance` and `PeriodicAllowance`, the latter allows you to setup a periodic allowance to another user. In this tutorial, you will set up two tokens in your blockchain: one default token called `stake`, which will be used for fees, and another called `kudos`, for sending to your friends.

Reviewer 3: ```suggestion
There are two [fee allowance types](https://docs.cosmos.network/v0.44/modules/feegrant/01_concepts.html#fee-allowance-types) that the feegrant module implements: `BasicAllowance` and `PeriodicAllowance`, the latter allows you to setup a periodic allowance to another user. In this tutorial, you will set up two tokens in your blockchain: one default token called `stake`, which will be used for fees, and another called `kudos`, for sending to your friends.
```

> An often discussed use case for the feegrant module is that blockchains or smart contracts can allow new users to start interacting with the chain or contract without having to acquire tokens first, greatly improving the onboarding experience.
> 
> There are two [fee allowance types](https://docs.cosmos.network/v0.44/modules/feegrant/01_concepts.html#fee-allowance-types) that the feegrant module implements: `BasicAllowance` and `PeriodicAllowance`, the latter allows you to setup a periodic allowance to another user. In this tutorial, you will set up two tokens in your blockchain: one default token called `stake`, which will be used for fees, and another called `kudos`, for sending to your friends.

Reviewer 2: ```suggestion
An often discussed use case for the fee grant module is improved onboarding experience because new users don't have to acquire tokens before they can start interacting with the blockchain or smart contract.  

Two [fee allowance types](https://docs.cosmos.network/v0.44/modules/feegrant/01_concepts.html#fee-allowance-types) are implemented with the fee grant module:

- `BasicAllowance`

    Grantee uses fees from a granter's account.

- `PeriodicAllowance`

    A periodic allowance to another user. 

In this tutorial, you will set up two tokens in your blockchain: a default token called `stake` to use for fees and another token called `kudos` to send to your friends.
```

Reviewer 2: let's see if using list items works here for the two allowance types.

> You will learn how to spin up a single node network using the simulation application in Cosmos SDK (`simapp`), let Charlie grant a `BasicAllowance` to Bob, and then have Bob send `kudos` tokens to Charlie, while having zero `stake` to pay for fees.

Reviewer 3: Not sure if you want to have this in the repo: when _creating_ users, A, B, C and so on we try to keep the names feminine / masculine / fem / masc and so on. That would mean Alice, Bob, Carol (or other feminine name), Dave (or other masculine name)...

Reviewer 3: Additionally, to be predictably conventional, I would introduce the names in order:

> ...single node network ... validated by **Alice**, let **Bob** grant an allowance to **Carol** and then have Carol send...

Like this, no reader will think "Did I miss a mention of Alice earlier?"

Author: Makes total sense, thanks! I'll change it in the next commit.

> You will learn how to spin up a single node network using the simulation application in Cosmos SDK (`simapp`) and set up Alice to be a validator. Then you'll have Bob send `kudos` tokens to Carol, while having zero `stake` to pay for fees. The latter will be possible because you'll let Carol grant a `BasicAllowance` to Bob.

Reviewer 2: ```suggestion
- You will learn how to spin up a single node network using the simulation application in Cosmos SDK (`simapp`).
- Set Alice up to be a validator.
- Carol is the granter and grants a `BasicAllowance` to Bob.
- Bob is the grantee who receives a `BasicAllowance` that allows Bob to send `kudos` tokens to Carol, even though Bob has zero `stake` to pay for fees. 
```

Reviewer 2: I suggest that we avoid "the latter" and other positional references, let's r/w for clarity and simple word flow 

This tutorial overview is a good time to identify the role of each user

Author: This is much clearer, thanks! Changing the order of Bob and Carol according to @xavierlepretre feedback which mentioned the standard of introducing users alphabetically. 

```suggestion
- You will learn how to spin up a single node network using the simulation application in Cosmos SDK (`simapp`).
- Set Alice up to be a validator.
- Bob is the grantee who receives a `BasicAllowance` that allows Bob to send `kudos` tokens to Carol, even though Bob has zero `stake` to pay for fees. 
- Carol is the granter and grants a `BasicAllowance` to Bob.
```

## Requirements

Before you start the tutorial, you need to install the `simd` binary.

Clone the `cosmos-sdk` repository:

```bash
git clone https://github.com/cosmos/cosmos-sdk
```

Change directories and check out `v0.44.0`:

```bash
cd cosmos-sdk && git checkout v0.44.0
```

Install the `simd` binary:

```bash
make install
```

Check to make sure the installation was successful:

```bash
simd version
```

> You should see `0.44.0` printed to the console.

Reviewer 2: ```suggestion
The version output `0.44.0` is output to the console.
```

Reviewer 2: let's not use "should" see word list https://developers.google.com/style/word-list?hl=en#letter-s 

> should, should be
> Generally avoid.
> Because should is ambiguous by definition, it can be problematic. For example, if you're telling the reader what to do, should implies that the action is recommended but optional, leaving the reader unsure about what to do.
> Clarify what you mean.

Author: Thanks. How about this?

```suggestion
The version number `0.44.0` is output to the console.
```

## Configuration

<!-- TODO: update tip to use `unsafe-reset-all` with better user experience -->

::: tip
If you have used `simd` before, you might already have a `.simapp` directory in your home directory. To keep the previous data, either save the directory to another location or use the `--home` flag and specify a different directory for each command in the following instructions. If you don't want to keep the previous data, remove the previous directory (`rm -rf ~/.simapp`).
:::

Run the following commands to configure the `simd` binary.

Set the chain ID:

```bash
simd config chain-id demo
```

Set the [keyring backend](https://docs.cosmos.network/v0.42/run-node/keyring.html#the-test-backend):

```bash
simd config keyring-backend test
```

## Key Setup

You'll have to create a few test keys for your users.

> Add a key for Alice:

Reviewer 3: ```suggestion
Add a key for Alice, the validator:
```

```bash
simd keys add alice
```

> Add a key for Bob:

Reviewer 3: ```suggestion
Add a key for Bob, the grantee:
```

```bash
simd keys add bob
```

> And lastly, add a key for Charlie:

Reviewer 3: ```suggestion
And lastly, add a key for Charlie, the granter:
```

```bash
simd keys add carol
```

If you'd like to see an overview of your keys, use:

```bash
simd keys list
```

> In order to avoid having to copy-paste their address all the time, you should export them to variables that you can keep using during this tutorial:

Reviewer 2: ```suggestion
::: tip
To avoid having to copy and paste the user addresses now is a good time to export the user keys to variables that you can access and use for this tutorial. 
:::
```

```bash
export ALICE_KEY=$(simd keys show alice -a)
export BOB_KEY=$(simd keys show bob -a)
export CAROL_KEY=$(simd keys show carol -a)
```

## Chain Setup

The following commands set up a chain using the simulation application (`simapp`).

Initialize the node:

```bash
simd init test --chain-id demo
```

> Alice will be your validator. Add Alice and an initial balance to the genesis file:

Reviewer 2: ```suggestion
Alice is your validator. Add Alice and an initial balance to the genesis file:
```

```bash
simd add-genesis-account alice 5000000000stake --keyring-backend test
```

Add Bob and an initial balance to the genesis file:

```bash
simd add-genesis-account bob 2000kudos --keyring-backend test
```

> Note that Bob only has `kudos` tokens and won't be able to pay for any fees that might be needed.

Reviewer 2: ```suggestion
Note that Bob has only `kudos` tokens and won't be able to pay for any fees that might be needed.
```

> Add Carol and an initial balance to the genesis file:

Reviewer 2: ```suggestion
Add Carol and an initial `stake` balance to the genesis file:
```

```bash
simd add-genesis-account charlie 1000000stake --keyring-backend test
```

Generate a transaction to add Alice to the initial validator set:

```bash
simd gentx alice 1000000stake --chain-id demo
```

Add the validator transaction to the genesis file:

```bash
simd collect-gentxs
```

## Start Chain

You are now ready to start a single node network on your local machine.

Start the chain:

```bash
simd start
```

## Grant Allowance

> Before Bob can send `kudos` to Charlie, you'll need to setup an allowance for Bob, so that Charlie will pay for any gas fees the transaction might incur. You're not letting Alice pay the fees for Bob because Alice is the validator and would receive all those fees, which means you wouldn't be able to see the difference in her balance.

Reviewer 3: ```suggestion
Before Bob can send `kudos` to Charlie, you'll need to set up an allowance for Bob, so that Charlie will pay for any gas fees the transaction might incur. The exercise did not choose Alice as the granter because she is the validator too. Choosing Alice as both the validator and the granter would make the change in her balance as a result of her grantee's actions less evident.
```

Author: Not sure I agree with the wording here, but will make changes as I understand the way I phrased it might be confusing.

> Before Bob can send `kudos` to Carol, you'll need to set up an allowance for Bob, so that Carol will pay for any gas fees the transaction might incur. Note that Alice is the validator, so it would not make sense to make her the granter as well because a validator would receive all the fees that the granter paid for. In that case, you wouldn't have been able to see the difference in her balance after a transaction was completed.

Reviewer 2: ```suggestion
Before Bob can send `kudos` to Carol, you must set up an allowance for Bob so that Carol pays for any gas fees the transaction might incur. Because Alice is the validator, it does not make sense to make her the granter because a validator receives all the fees that the granter paid for. If Alice were the granter, there would not be a difference in her balance after a transaction was completed.
```

The `BasicAllowance` is a permission for a grantee to use up fees until the `spend_limit` or `expiration` is reached. Create an allowance with a spend limit of `100000stake` and no expiration date:

```bash
simd tx feegrant grant $CAROL_KEY $BOB_KEY --from carol --spend-limit 100000stake
```

View the allowance:

```bash
simd q feegrant grants $BOB_KEY
```

## Send Tokens

> First, let's check the balances of Alice, Bob and Carol, so that you can later confirm if your transaction was successful:

Reviewer 2: ```suggestion
First, let's check the balances of Alice, Bob, and Carol. Verifying the initial balance provides a baseline so that you can later confirm if your transaction was successful:
```

Reviewer 2: we love the serial comma https://developers.google.com/style/commas?hl=en

```bash
simd q bank balances $ALICE_KEY
simd q bank balances $BOB_KEY
simd q bank balances $CHARLIE_KEY
```

Note that Alice has `4999000000stake` because she bonded `1000000stake` to become a validator during the chain setup.

Any transaction that is sent using the `tx` command can be ammended with a flag `--fee-account` that takes an account as input which will pay for the fees.

> Send `kudos` tokens from Bob to Charlie, while Charlie pays fees:

Reviewer 3: ```suggestion
Send `kudos` tokens from Bob to Charlie, while Charlie pays the fees:
```

> Any transaction that is sent using the `tx` command can be ammended with a flag `--fee-account` that takes an account as input which will pay for the fees.

Reviewer 2: ```suggestion
Any transaction that is sent using the `tx` command can use the `--fee-account` flag to specify an account as input to pay for the fees.
```

Reviewer 2: hmmm I think "amended" is an incorrect word choice here, as amended means to change after the face, can we instead use this rewrite?

```bash
simd tx bank send $BOB_KEY $CAROL_KEY 100kudos --from bob --fee-account $CAROL_KEY --fees 500stake
```

Look at the balances again:

```bash
simd q bank balances $ALICE_KEY
simd q bank balances $BOB_KEY
simd q bank balances $CAROL_KEY
```

> Notice how Carol lost the `500stake` that we added to the transaction, while Bob was the one who signed.

Reviewer 2: ```suggestion
Notice how Carol no longer has the `500stake` that was added to the transaction that Bob signed.
```

Reviewer 2: Carol didn't "lose" the fees... :)

View the allowance again:

```bash
simd q feegrant grants $BOB_KEY
```

Note how `spend_limit` has been reduced and Bob now has 99500stake left to spend on fees.

## Revoke Allowance

The granter can revoke the allowance from the grantee using the `revoke` command.

Revoke allowance:

```bash
simd tx feegrant revoke $CAROL_KEY $BOB_KEY --from carol
```

View the allowance:

```bash
simd q feegrant grants $BOB_KEY
```

## 🎉 Congratulations 🎉

By completing this tutorial, you have learned how to use the feegrant module:

- Configured and used the simulation application (simapp)
- Created an allowance
- Made a transaction with fees paid by a granter
- Revoked an allowance

> There is a lot more that you can do with the feegrants module. You can add a list of allowed messages, set an expiration date and set a time duration after which the spend limit is refilled again. To learn more about the feegrant module and different types of allowances, check out the [feegrant module documentation](https://docs.cosmos.network/v0.44/modules/feegrant/).

Reviewer 2: ```suggestion
There is a lot more that you can do with the fee grants module. You can add a list of allowed messages, set an expiration date, and set a time duration after which the spend limit is refilled. To learn more about the fee grant module and different types of allowances, check out the Cosmos SDK [Fee grant module](https://docs.cosmos.network/v0.44/modules/feegrant/) documentation.
```

export ALICE_KEY=$(simd keys show alice -a)
export BOB_KEY=$(simd keys show bob -a)
export CAROL_KEY=$(simd keys show carol -a)
```

## Chain Setup

The following commands set up a chain using the simulation application (`simapp`).

Initialize the node:

```bash
simd init test --chain-id demo
```

Alice is your validator. Add Alice and an initial balance to the genesis file:

```bash
simd add-genesis-account alice 5000000000stake --keyring-backend test
```

Add Bob and an initial balance to the genesis file:

```bash
simd add-genesis-account bob 2000kudos --keyring-backend test
```

Note that Bob has only `kudos` tokens and won't be able to pay for any fees that might be needed.

Add Carol and an initial `stake` balance to the genesis file:

```bash
simd add-genesis-account carol 1000000stake --keyring-backend test
```

Generate a transaction to add Alice to the initial validator set:

```bash
simd gentx alice 1000000stake --chain-id demo
```

Add the validator transaction to the genesis file:

```bash
simd collect-gentxs
```

## Start Chain

You are now ready to start a single node network on your local machine.

Start the chain:

```bash
simd start
```

## Grant Allowance

> Before Bob can send `kudos` to Carol, you must set up an allowance for Bob so that Carol pays for any gas fees the transaction might incur. Because Alice is the validator, it does not make sense to make her the granter because a validator receives all the fees that the granter paid for. If Alice were the granter, there would not be a difference in her balance after a transaction was completed.

Reviewer: Alice does not directly receive the fees from the transaction. For example, when you run the balance query a second time after executing the send transaction, Alice's balance is unchanged.

Validators must claim their rewards through a separate transaction. Also, block rewards are distributed proportionally to all validators relative to their voting power. Yes, Alice is the only validator in this case but this statement might cause some confusion (i.e. "a validator receives all the fees").

I think this statement is a bit confusing and the example could in fact be done with only two accounts, one account being both the validator and granter (or grantee). I think three accounts works too and the separation might help beginners but I would remove or revise the second part of this paragraph as it might cause confusion.

```suggestion
Before Bob can send `kudos` to Carol, you must set up an allowance for Bob so that Carol pays for any gas fees the transaction might incur.
```

Author: AH! That's a really good point @ryanchristo , thanks for highlighting! This does make me wonder whether or not to just stick to two accounts and reduce the complexity. I think there's no reason to stick to three accounts really. Wdyt?

Reviewer: Up to you! I probably would have done two but three works too.

simd q bank balances $BOB_KEY
simd q bank balances $CAROL_KEY
```

Note that Alice has `4999000000stake` because she bonded `1000000stake` to become a validator during the chain setup.

Any transaction that is sent using the `tx` command can use the `--fee-account` flag to specify an account as input to pay for the fees.

Send `kudos` tokens from Bob to Carol, while Carol pays the fees:

```bash
simd tx bank send $BOB_KEY $CAROL_KEY 100kudos --from bob --fee-account $CAROL_KEY --fees 500stake
```

Look at the balances again:

```bash
simd q bank balances $ALICE_KEY
simd q bank balances $BOB_KEY
simd q bank balances $CAROL_KEY
```

Notice how Carol no longer has the `500stake` that was added to the transaction that Bob signed.

View the allowance again:

```bash
simd q feegrant grants $BOB_KEY
```

Note how `spend_limit` has been reduced and Bob now has 99500stake left to spend on fees.

## Revoke Allowance

The granter can revoke the allowance from the grantee using the `revoke` command.

Revoke allowance:

```bash
simd tx feegrant revoke $CAROL_KEY $BOB_KEY --from carol
```

View the allowance:

```bash
simd q feegrant grants $BOB_KEY
```

## 🎉 Congratulations 🎉

> By completing this tutorial, you have learned how to use the feegrant module:

Reviewer: Another one for consistency.

```suggestion
By completing this tutorial, you have learned how to use the fee grant module:
```

Author: Thanks Ryan! I guess this is another question then: are we sticking with capitals when referring to the module?

```suggestion
By completing this tutorial, you have learned how to use the Fee Grant module:
```

Reviewer: I would have written as "feegrant" so not entirely sure. I think consistency is the most important thing so if capitalized here, just make sure you do so elsewhere. Maybe @barriebyron could weigh in?

- Configured and used the simulation application (simapp)
- Created an allowance
- Made a transaction with fees paid by a granter
- Revoked an allowance

> There is a lot more that you can do with the fee grants module. You can add a list of allowed messages, set an expiration date, and set a time duration after which the spend limit is refilled. To learn more about the fee grant module and different types of allowances, check out the Cosmos SDK [Fee grant module](https://docs.cosmos.network/v0.44/modules/feegrant/) documentation.

Reviewer: ```suggestion
There is a lot more that you can do with the fee grant module. You can add a list of allowed messages, set an expiration date, and set a time duration after which the spend limit is refilled. To learn more about the fee grant module and different types of allowances, check out the Cosmos SDK [Fee grant module](https://docs.cosmos.network/v0.44/modules/feegrant/) documentation.
```

Author: Same question as above.

```suggestion
There is a lot more that you can do with the Fee Grant module. You can add a list of allowed messages, set an expiration date, and set a time duration after which the spend limit is refilled. To learn more about the fee grant module and different types of allowances, check out the Cosmos SDK [Fee grant module](https://docs.cosmos.network/v0.44/modules/feegrant/) documentation.
```
